# Census 2011: Population, Literacy and Workforce Across Indian States

## Data Source
Primary Census Abstract 2011, Census of India,
Registrar General & Census Commissioner, India.

In [6]:
!pip install xlrd

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [8]:
import pandas as pd


In [9]:
df = pd.read_excel("PCA0000_2011_MDDS.xls")



In [10]:
df.shape

(108, 98)

In [11]:
df.columns

Index(['State Code', 'District Code', 'Subdistt Code', 'Town/Village Code',
       'State Code.1', 'District Code.1', 'Subdistt Code.1',
       'Town/Village Code.1', 'Ward Code', 'EB Code', 'Level', 'Name', 'TRU',
       'No of Households', 'Total Population Person', 'Total Population Male',
       'Total Population Female', 'Population in the age group 0-6 Person',
       'Population in the age group 0-6 Male',
       'Population in the age group 0-6 Female',
       'Scheduled Castes population Person',
       'Scheduled Castes population Male', 'Scheduled Castes populationFemale',
       'Scheduled Tribes population Person',
       'Scheduled Tribes population Male',
       'Scheduled Tribes population Female', 'Literates Population Person',
       'Literates Population Male', 'Literates Population Female',
       'Illiterate Persons', 'Illiterate Male', 'Illiterate Female',
       'Total Worker Population Person', 'Total Worker Population Male',
       'Total Worker Population Fema

In [12]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 98 columns):
 #   Column                                                Non-Null Count  Dtype
---  ------                                                --------------  -----
 0   State Code                                            108 non-null    int64
 1   District Code                                         108 non-null    int64
 2   Subdistt Code                                         108 non-null    int64
 3   Town/Village Code                                     108 non-null    int64
 4   State Code.1                                          108 non-null    int64
 5   District Code.1                                       108 non-null    int64
 6   Subdistt Code.1                                       108 non-null    int64
 7   Town/Village Code.1                                   108 non-null    int64
 8   Ward Code                                             108 non-null    int64
 9   EB Code   

#Check missing values   

In [13]:
df.isnull().sum()

State Code                                      0
District Code                                   0
Subdistt Code                                   0
Town/Village Code                               0
State Code.1                                    0
                                               ..
Marginal Other Workers Population 0_3 Male      0
Marginal Other Workers Population 0_3 Female    0
Non Working Population Person                   0
Non Working Population Male                     0
Non Working Population Female                   0
Length: 98, dtype: int64

In [14]:
df.duplicated().sum()

np.int64(0)

#Let's find 10 states with the highest Marginal Worker Population Female

In [20]:


top_10_female_marginal = df.sort_values(
    "Marginal Worker Population Female",
    ascending=False
).head(10)

top_10_female_marginal[
    ["Name", "Marginal Worker Population Female"]
]

,Name,Marginal Worker Population Female
0,India,60580320
1,India,54119910
27,UTTAR PRADESH,8752760
28,UTTAR PRADESH,7871631
2,India,6460410
24,RAJASTHAN,5774748
25,RAJASTHAN,5506258
30,BIHAR,5413877
31,BIHAR,5187196
69,MADHYA PRADESH,5087109


## 6. Derived indicators
- **Sex ratio** = females per 1000 males
- **Literacy rate (%)** = literates / total population * 100
- **Work participation rate (%)** = total workers / total population * 100

#This code filters the data down to one row per state (combined rural+urban totals), calculates three indicators for each state — sex ratio, literacy rate, and work participation rate — then sorts states by population and shows the top 10 largest states along with these indicators.

In [22]:
states = df[(df['Level'] == 'STATE') & (df['TRU'] == 'Total')].copy().reset_index(drop=True)

states['Sex Ratio (F per 1000 M)'] = (
    states['Total Population Female'] / states['Total Population Male'] * 1000
).round(1)
states['Literacy Rate (%)'] = (
    states['Literates Population Person'] / states['Total Population Person'] * 100
).round(2)
states['Work Participation Rate (%)'] = (
    states['Total Worker Population Person'] / states['Total Population Person'] * 100
).round(2)

states.sort_values('Total Population Person', ascending=False).head(10)



,State Code,District Code,Subdistt Code,Town/Village Code,State Code.1,District Code.1,Subdistt Code.1,Town/Village Code.1,Ward Code,EB Code,...,Marginal Household Industries Population 0_3 Female,Marginal Other Workers Population 0_3 Person,Marginal Other Workers Population 0_3 Male,Marginal Other Workers Population 0_3 Female,Non Working Population Person,Non Working Population Male,Non Working Population Female,Sex Ratio (F per 1000 M),Literacy Rate (%),Work Participation Rate (%)
8,9,0,0,0,9,0,0,0,0,0,...,210563,963558,621353,342205,133997626,54633748,79363878,912.4,57.25,32.94
26,27,0,0,0,27,0,0,0,0,0,...,35996,286328,176608,109720,62946455,25626181,37320274,929.4,72.57,43.99
9,10,0,0,0,10,0,0,0,0,0,...,105889,342975,212562,130413,69374465,29055968,40318497,917.9,50.44,33.36
18,19,0,0,0,19,0,0,0,0,0,...,176310,556465,320266,236199,56519760,20092980,36426780,950.0,67.42,38.08
27,28,0,0,0,28,0,0,0,0,0,...,34197,248457,137424,111033,45157871,18256551,26901320,992.8,59.77,46.61
22,23,0,0,0,23,0,0,0,0,0,...,43710,266045,149020,117025,41052676,17465336,23587340,930.9,59.00,43.47
32,33,0,0,0,33,0,0,0,0,0,...,41039,287895,165007,122888,39262349,14702997,24559352,996.4,71.85,45.58
7,8,0,0,0,8,0,0,0,0,0,...,32229,445760,163693,282067,38662182,17253921,21408261,928.2,55.84,43.60
28,29,0,0,0,29,0,0,0,0,0,...,25504,269255,149118,120137,33222700,12696541,20526159,972.9,66.53,45.62
23,24,0,0,0,24,0,0,0,0,0,...,9546,137026,73149,63877,35671945,13490346,22181599,919.3,67.99,40.98


In [ ]:
#Top and bottom states by key indicators
This code prints four short rankings: the 5 most populous states, the 5 states with the highest literacy rate, the 5 states with the lowest literacy rate, and the 5 states with the highest sex ratio — each shown with just the state name and the relevant value.

In [23]:
print("Top 5 most populous states:")
display(states.nlargest(5, 'Total Population Person')[['Name', 'Total Population Person']])

print("\nTop 5 states by literacy rate:")
display(states.nlargest(5, 'Literacy Rate (%)')[['Name', 'Literacy Rate (%)']])

print("\nBottom 5 states by literacy rate:")
display(states.nsmallest(5, 'Literacy Rate (%)')[['Name', 'Literacy Rate (%)']])

print("\nTop 5 states by sex ratio:")
display(states.nlargest(5, 'Sex Ratio (F per 1000 M)')[['Name', 'Sex Ratio (F per 1000 M)']])

Top 5 most populous states:


,Name,Total Population Person
8,UTTAR PRADESH,199812341
26,MAHARASHTRA,112374333
9,BIHAR,104099452
18,WEST BENGAL,91276115
27,ANDHRA PRADESH,84580777



Top 5 states by literacy rate:


,Name,Literacy Rate (%)
31,KERALA,84.22
30,LAKSHADWEEP,81.51
29,GOA,79.91
24,DAMAN & DIU,77.45
34,ANDAMAN & NICOBAR ISLANDS,77.32



Bottom 5 states by literacy rate:


,Name,Literacy Rate (%)
9,BIHAR,50.44
11,ARUNACHAL PRADESH,55.36
19,JHARKHAND,55.56
7,RAJASTHAN,55.84
0,JAMMU & KASHMIR,56.35



Top 5 states by sex ratio:


,Name,Sex Ratio (F per 1000 M)
31,KERALA,1084.3
33,PUDUCHERRY,1037.4
32,TAMIL NADU,996.4
27,ANDHRA PRADESH,992.8
13,MANIPUR,992.3


#SC/ST population share

In [24]:
states['SC Population (%)'] = (states['Scheduled Castes population Person'] / states['Total Population Person'] * 100).round(2)
states['ST Population (%)'] = (states['Scheduled Tribes population Person'] / states['Total Population Person'] * 100).round(2)
states.nlargest(5, 'ST Population (%)')[['Name', 'ST Population (%)']]

,Name,ST Population (%)
30,LAKSHADWEEP,94.80
14,MIZORAM,94.43
12,NAGALAND,86.48
16,MEGHALAYA,86.15
11,ARUNACHAL PRADESH,68.79


#Gender gap in literacy

In [25]:
states['Male Literacy (%)'] = (states['Literates Population Male'] / states['Total Population Male'] * 100).round(2)
states['Female Literacy (%)'] = (states['Literates Population Female'] / states['Total Population Female'] * 100).round(2)
states['Literacy Gender Gap'] = (states['Male Literacy (%)'] - states['Female Literacy (%)']).round(2)
states.nlargest(5, 'Literacy Gender Gap')[['Name', 'Male Literacy (%)', 'Female Literacy (%)', 'Literacy Gender Gap']]

,Name,Male Literacy (%),Female Literacy (%),Literacy Gender Gap
7,RAJASTHAN,66.63,44.21,22.42
25,DADRA & NAGAR HAVELI,73.56,53.82,19.74
19,JHARKHAND,64.28,46.37,17.91
21,CHHATTISGARH,68.64,51.70,16.94
8,UTTAR PRADESH,65.31,48.42,16.89


#Household size (avg persons per household)

In [27]:
states['Avg Household Size'] = (states['Total Population Person'] / states['No of Households']).round(2)
states.nlargest(5, 'Avg Household Size')[['Name', 'Avg Household Size']]

,Name,Avg Household Size
8,UTTAR PRADESH,5.97
0,JAMMU & KASHMIR,5.92
30,LAKSHADWEEP,5.57
9,BIHAR,5.50
16,MEGHALAYA,5.41


#Child population share (0–6 age group) as a proxy for future demographic trends

In [28]:
states['Child Population Share (%)'] = (states['Population in the age group 0-6 Person'] / states['Total Population Person'] * 100).round(2)
states.nlargest(5, 'Child Population Share (%)')[['Name', 'Child Population Share (%)']]

,Name,Child Population Share (%)
16,MEGHALAYA,19.16
9,BIHAR,18.38
19,JHARKHAND,16.34
0,JAMMU & KASHMIR,16.10
7,RAJASTHAN,15.54


#Regional grouping (North/South/East/West/Northeast) using a mapping dict, then group-by comparisons

In [30]:
region_map = {
    'JAMMU & KASHMIR': 'North', 'HIMACHAL PRADESH': 'North', 'PUNJAB': 'North',
    'CHANDIGARH': 'North', 'UTTARAKHAND': 'North', 'HARYANA': 'North',
    'NCT OF DELHI': 'North', 'RAJASTHAN': 'North', 'UTTAR PRADESH': 'North',

    'BIHAR': 'East', 'JHARKHAND': 'East', 'ODISHA': 'East', 'WEST BENGAL': 'East',

    'ASSAM': 'Northeast', 'ARUNACHAL PRADESH': 'Northeast', 'MANIPUR': 'Northeast',
    'MEGHALAYA': 'Northeast', 'MIZORAM': 'Northeast', 'NAGALAND': 'Northeast',
    'SIKKIM': 'Northeast', 'TRIPURA': 'Northeast',

    'MADHYA PRADESH': 'Central', 'CHHATTISGARH': 'Central',

    'GUJARAT': 'West', 'MAHARASHTRA': 'West', 'GOA': 'West',
    'DADRA & NAGAR HAVELI': 'West', 'DAMAN & DIU': 'West',

    'ANDHRA PRADESH': 'South', 'KARNATAKA': 'South', 'KERALA': 'South',
    'TAMIL NADU': 'South', 'PUDUCHERRY': 'South',
    'ANDAMAN & NICOBAR ISLANDS': 'South', 'LAKSHADWEEP': 'South'
}

states['Region'] = states['Name'].map(region_map)
states.groupby('Region')[['Literacy Rate (%)', 'Sex Ratio (F per 1000 M)']].mean().round(2)

,Literacy Rate (%),Sex Ratio (F per 1000 M)
Region,,
Central,59.60,960.75
East,59.28,948.80
North,66.24,902.59
Northeast,67.52,954.21
South,73.99,986.61
West,72.57,842.86


## Story Pitch: "The Kerala Outlier"

**The hook:** In a country where men outnumber women in almost every major state, one state flips the script entirely.

**The story:** Kerala is the only major Indian state where women outnumber men (1,084 females per every 1,000 males), and it also tops the country in literacy at 84.2% — nearly 34 percentage points ahead of Bihar, the lowest-ranked major state (50.4%). This isn't a random data point; decades of research point to Kerala's investment in female education, healthcare access, and social welfare as the reason both numbers moved together.
 As India debates its next Census and updates to social welfare policy, Kerala's decades-old outcomes offer a real, measurable benchmark for what's achievable.